go/vt/mysqlctl/compression.go

```diff
@@ -0,0 +1,254 @@
+package mysqlctl
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os/exec"
+	"strings"
+	"sync"
+
+	"github.com/klauspost/compress/zstd"
+	"github.com/klauspost/pgzip"
+	"github.com/pierrec/lz4"
+	"github.com/planetscale/pargzip"
+
+	"vitess.io/vitess/go/vt/logutil"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+var (
+	compressionLevel = flag.Int("compression_level", 1, "What level to pass to the compressor")
+
+	errUnsupportedCompressionEngine    = errors.New("unsupported engine")
+	errUnsupportedCompressionExtension = errors.New("unsupported extension")
+
+	// this is use by getEngineFromExtension() to figure out which engine to use in case the user didn't specify
+	engineExtensions = map[string][]string{
+		".gz":  {"pgzip", "pargzip"},
+		".lz4": {"lz4"},
+		".zst": {"zstd"},
+	}
+)
+
+func getEngineFromExtension(extension string) (string, error) {
+	for ext, eng := range engineExtensions {
+		if ext == extension {
+			return eng[0], nil // we select the first supported engine in auto mode
+		}
+	}
+
+	return "", fmt.Errorf("%w \"%s\"", errUnsupportedCompressionExtension, extension)
```

Comment:
Anywhere where you have `\"%s\"` in a format string, you can replace it with `%q`, which adds the quotes automatically _and_ escapes the contents of the string so they always look proper inside the quotes. :ok_hand:

Reply:
neat, I will change that!